lib/project_types/extension/models/specifications.rb

```diff
@@ -0,0 +1,61 @@
+module Extension
+  module Models
+    class Specifications
+      include SmartProperties
+
+      property! :custom_handler_root,
+        accepts: ->(d) { File.directory?(d) },
+        default: -> { File.expand_path('lib/project_types/extension/models/specification_handlers', ShopifyCli::ROOT) }
+
+      property! :custom_handler_namespace,
+        accepts: ->(m) { m.respond_to?(:const_get) },
+        default: -> { ::Extension::Models::SpecificationHandlers }
+
+      property! :fetch_specifications,
+        accepts: ->(p) { p.respond_to?(:to_proc) }
+
+      def initialize(*)
+        super
+        @handlers = fetch_specifications_and_build_handlers
+      end
+
+      def [](identifier)
+        handlers[identifier]
+      end
+
+      def valid?(identifier)
+        handlers.key?(identifier)
+      end
+
+      def each(&block)
+        handlers.values.each(&block)
+      end
+
+      protected
+
+      attr_reader :handlers
+
+      private
+
+      def fetch_specifications_and_build_handlers
+        ShopifyCli::Result
+          .wrap(&fetch_specifications)
+          .call
```

Comment:
I think if I am reading this right we make the `.call` separately because we are deferring the execution of the `fetch_specifications`. However in this case we are never actually doing anything by differing the call since we aren't passing out the deferred call to allow a different call site to decide when to execute. Should we just simplify this by calling the `fetch_specifications` directly within the `wrap`?

Reply:
Technically we could simply do `fetch_specifications.call` as the method object returns a result. However, using Result.wrap (which won't double wrap) adds support for passing a lambda expression, which is handy in tests. I could have used, 

```ruby
Result.wrap(fetch_specifications.call)
```

but that wouldn't protect against exceptions thrown by `fetch_specifications`, which can occur if `fetch_specifications` is a lambda expression rather than a method object that guards against this case by returning a failure result. I decided one extra line of code for a more flexible interface is an ok tradeoff to make. If you think this makes the code harder to understand, I can simply go with `fetch_specifications.call`.